### ISSUE TITLE
Multi-thread training story: HOGWILD! + in-process DDP + striped-lock trainers (exceed PyTorch DDP)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1369]
## Goal — exceed PyTorch / TensorFlow on multi-thread training

PyTorch and TensorFlow are fundamentally single-thread-per-instance at the
`nn.Module.forward` / `tf.keras.Model.__call__` level. To train in parallel
they require:

- **`torch.distributed`** with one OS process per worker and one model
  replica per process, gluing them together with NCCL/Gloo all-reduce.
- **`torch.nn.DataParallel`**, which is deprecated, has the well-documented
  Python-GIL bottleneck, and still serializes the parameter-server step.
- **HOGWILD!** (Niu et al. 2011), which is supported by PyTorch only via
  `torch.multiprocessing` — again, one process per worker because
  per-instance state isn't thread-safe.

That means a 64-core CPU box can only run one training thread per model.
We can beat this by making the per-instance state lock-free THREAD-safe
(not just process-safe), and shipping HOGWILD!-style and DDP-style
trainers that run in-process across `Thread`s — no `multiprocessing`,
no GIL, no inter-process IPC tax, no per-replica model copy.

This issue lays out the design.

---

## Phase 1 (DONE in #1362) — make per-instance optimizer state thread-safe

All 25 per-tensor `Dictionary<Tensor<T>, ...>` instance fields across
the 15 stateful optimizers (Adam / AdamW / Adam8Bit / AMSGrad / AdaDelta /
AdaMax / Adagrad / FTRL / LAMB / LARS / Lion / Momentum / Nadam /
Nesterov / RMSProp) plus `MixedPrecisionContext._fp32SnapshotPool`
have been swapped to `ConcurrentDictionary`.

**Concurrency contract for callers**: lock-free reads via
`TryGetValue`; atomic upsert via `AddOrUpdate`; **never** call `.Count`
or `.IsEmpty` on these — they acquire per-bucket `Monitor` locks and
serialize parallel readers (root-caused 2026-04-22 in
`DeferredArrayMaterializer` where high-fanout parallel tensor forwards
observed 44 s of unmanaged wait per 30 s of work). If a "populated?"
indicator is needed, add a separate `Interlocked.Increment`-driven
volatile counter alongside the dictionary (see
`DeferredArrayMaterializer._pendingCount`).

In addition, `NeuralNetworkBase.TrainWithTape` now carries an
`Interlocked.CompareExchange` sentinel that throws with a clear redirect
if two threads enter on the same instance — until the Phase 2/3 APIs
below land, callers MUST shard the network per worker.

---

## Phase 2 — HOGWILD!-style lock-free trainer

**Reference**: Recht, Re, Wright, Niu (2011), "HOGWILD!: A Lock-Free
Approach to Parallelizing Stochastic Gradient Descent".

**Key insight**: when gradients are sparse (most training steps touch
only a small fraction of parameters), the race conditions in unlocked
parameter writes are statistically negligible. The model still
converges, often FASTER than locked variants because of cache-line
sharing and reduced contention.

**Concrete API surface**:

```csharp
// One model instance. N worker threads. All workers write through to
// the SAME parameter tensors. Adam m/v are ConcurrentDictionary-backed
// so updates are atomic-per-key. No per-shard parameter copy, no
// all-reduce.
var trainer = new HogwildTrainer<float, Tensor<float>, Tensor<float>>(
    model,
    workerCount: Environment.ProcessorCount,
    options: new HogwildOptions {
        // HOGWILD! converges as long as gradient overlap is below this
        // theoretical threshold (Recht et al §3). Tighten on dense
        // workloads, loosen on sparse.
        MaxStaleness = 4,
        // If true, race-detection asserts run in DEBUG only: catches
        // pathological overlap that would prevent convergence.
        TraceRaceDetection = false,
    });
await trainer.TrainAsync(dataset, epochs: 10);
```

Internal mechanism: each worker calls `model.TrainStep(batch_i)` (a NEW
public API that BYPASSES the reentrancy sentinel). The sentinel is
replaced by the HOGWILD! contract — all per-tensor state must be
`ConcurrentDictionary`-backed (Phase 1 done) and all parameter tensor
writes must use `Volatile.Write` / `Interlocked.CompareExchange` on the
underlying `float*` (Tensors-side fused kernel work — separate issue).

**Expected speedup**: linear up to ~16 workers on x86-64 Skylake-class
chips; near-linear up to socket count on AMD EPYC / Intel Sapphire
Rapids. Sublinear scaling past that point is dominated by cache-line
ping-pong on the densest layers (LM-head embedding row updates).

**Tensors-side requirement**: fused-Adam kernels (Tensors `Sgemm`,
`AddElementWise`, etc.) must emit `volatile`/`atomic` stores when
inside a HOGWILD! step. Track in a separate Tensors issue.

---

## Phase 3 — DDP-style sharded trainer

When the model is too big to fit one copy in RAM and you DO want
all-reduce semantics (i.e. determinism, reproducibility, exact-Adam
math), provide `DistributedDataParallelTrainer`:

```csharp
var trainer = new DistributedDataParallelTrainer<float, Tensor<float>, Tensor<float>>(
    modelFactory: () => BuildModel(),    // one replica per shard
    shardCount: Environment.ProcessorCount,
    options: new DDPOptions {
        // "All-reduce after every step" — strictest determinism.
        // "Local SGD" — N local steps then sync — better throughput.
        Strategy = DDPStrategy.AllReduceEveryStep,
        // Ring vs hierarchical vs tree all-reduce. Ring is bandwidth-
        // optimal on a single NUMA node; hierarchical wins across nodes.
        AllReduceTopology = AllReduceTopology.Ring,
    });
await trainer.TrainAsync(dataset, epochs: 10);
```

Internal mechanism: each shard has its own model + own optimizer state
(no `ConcurrentDictionary` needed because shards never share state).
After every step (or every K steps in Local SGD), the shards swap
gradients via an in-process all-reduce ring buffer. No process
boundary, no IPC, no NCCL dependency.

**Why this is better than `torch.distributed`**: no separate process,
no GIL, no NCCL/Gloo dependency, sub-microsecond ring-buffer all-reduce
(vs. PyTorch's ~50 µs minimum for inter-process all-reduce).

**Determinism contract**: `AllReduceEveryStep` + fixed worker count +
fixed seed = bit-exact reproducibility across runs.

---

## Phase 4 — per-param locks (rarely needed but free to ship)

For users who want the SAFETY of HOGWILD! but with strict-correctness
guarantees instead of HOGWILD!'s probabilistic convergence, ship
`StripedLockTrainer`:

```csharp
// One striped Monitor lock per ~1024 parameters. Workers compete for
// the lock on the stripe they're writing. Strict correctness at the
// cost of ~5-15% throughput vs HOGWILD!.
var trainer = new StripedLockTrainer<float, Tensor<float>, Tensor<float>>(
    model, workerCount: Environment.ProcessorCount,
    options: new StripedLockOptions { StripeSize = 1024 });
```

---

## Phase 5 — benchmarks vs PyTorch DDP

After Phase 2 + 3 land, add a benchmark suite that compares:

| Scenario | PyTorch DDP (2 procs, NCCL) | AiDotNet HogwildTrainer (2 threads) | AiDotNet DDPTrainer (2 shards) |
|---|---|---|---|
| MLP-3, Iris | TBD | TBD | TBD |
| Transformer-L=6, WikiText-2 | TBD | TBD | TBD |
| ResNet-18, CIFAR-10 | TBD | TBD | TBD |
| LSTM-2, Penn Treebank | TBD | TBD | TBD |

Targets:
- 3x walltime improvement over PyTorch DDP on small CPU models (where
  PyTorch's inter-process overhead dominates the actual gradient work).
- 1.2x walltime improvement on medium-sized CPU models.
- Match PyTorch on GPU-bound models (where gradient compute dominates).
- Bit-exact match between `DDPTrainer(AllReduceEveryStep)` and the
  serial-equivalent reference (correctness regression gate).

---

## Phase 6 — public-facing API on AiModelBuilder

```csharp
var model = await new AiModelBuilder<float, Tensor<float>, Tensor<float>>()
    .ConfigureModelType(ModelType.NeuralNetwork)
    .ConfigureHogwildTraining(new HogwildOptions {
        WorkerCount = Environment.ProcessorCount,
    })
    .ConfigureOptimizer(new AdamOptimizer<float, Tensor<float>, Tensor<float>>())
    .BuildAsync();
```

`ConfigureHogwildTraining`, `ConfigureDDPTraining`, and
`ConfigureStripedLockTraining` are mutually exclusive — calling more
than one throws at the second call. `BuildAsync` plumbs the chosen
strategy into the trainer wrapper around the inner model.

---

## Open design questions

1. **Mixed precision under HOGWILD!**: the `_fp32SnapshotPool` is now
   `ConcurrentDictionary`, but the round-trip in `TrainWithTape`'s MP
   path reads the FP32 snapshot, casts to BF16, runs forward / backward,
   then casts back. Under HOGWILD! two workers could observe interleaved
   master / working weights. Is the resulting numeric noise within
   HOGWILD!'s convergence envelope? Need experimental validation.
2. **Gradient checkpointing under HOGWILD!**: AiDotNet's gradient
   checkpointing recomputes forward activations in backward. Under
   HOGWILD! the recompute could pick up another worker's mid-write
   parameters. May need explicit "checkpoint snapshot" — at gradient
   tape build time, snapshot the params seen by forward, recompute from
   that snapshot. Costs ~1x param memory.
3. **CUDA / DirectGpu compatibility**: HOGWILD! on GPU is non-trivial —
   it requires either CUDA kernels that emit `atomicCAS` for parameter
   writes (slow) or per-shard model copies (defeats HOGWILD!'s no-copy
   advantage). DDPTrainer is the better fit for GPU. Worth gating
   `ConfigureHogwildTraining` to CPU-only initially?
4. **Replay determinism**: ring-buffer all-reduce in Phase 3 needs to
   be deterministic given the same seed + worker count. The
   ring-traversal order must be fixed across runs — schedule shards by
   ID, not by wall-clock arrival.

---

## Acceptance criteria for the multi-thread training story

- [ ] Phase 1 ConcurrentDictionary sweep merged (#1362)
- [ ] `HogwildTrainer<T,...>` class lands with unit tests (single-thread
      equivalence, race-detection asserts in DEBUG)
- [ ] `DistributedDataParallelTrainer<T,...>` class lands with
      bit-exact-match regression test
- [ ] `AiModelBuilder.ConfigureHogwildTraining` / `ConfigureDDPTraining`
      facade methods exposed
- [ ] Benchmark suite vs PyTorch DDP on 4 reference models (Phase 5)
- [ ] Docs: `docs/multi-thread-training.md` covering the three trainers,
      when to use which, the HOGWILD! convergence contract, and the
      Tensors-side atomic-store requirement.

**Comment by ooples** [comment]
Triage note (PR #1789 sweep): reviewed against the #1789 branch. There is related activity on that branch but **no verified fix for what this issue actually asks for**, so it is deliberately left open and not linked for auto-close. Re-check after #1789 merges.

**Comment by ooples** [comment]
### Re-audit against master (`origin/master` @ `a22d7e78bf` (2026-09-05))

**Why this correction exists.** The `Triage note (PR #1789 sweep)` comment on this issue was produced by checking the **#1789 branch**, which was cut on **2026-07-05**. That sweep ran on **2026-08-14**, the day before #1789 merged. Any fix that reached `master` in that 5.5-week window *outside* #1789 was invisible to it, so those verdicts are false-negative-prone. Re-verified below against master itself.

**Verdict: OPEN - not started, correctly stated**

Confirmed: no HOGWILD!, in-process DDP, or striped-lock trainer implementation exists on master. The issue is accurate as written.

Triaged v1 Tier 3 (post-v1 epic).

---
*Posted by an automated re-audit of all 45 open issues against master on 2026-09-05. Every figure above was measured on the tree or read from CI run artifacts; none is carried over from a previous comment.*
